shared/pubsub.js

```diff
@@ -0,0 +1,579 @@
+'use strict'
+
+import type { JSONObject, JSONType } from '~/shared/types.js'
+
+// ====== Types ====== //
+
+/*
+ * Flowtype usage notes:
+ *
+ * - The '+' prefix indicates properties that should not be re-assigned or
+ *   deleted after their initialization.
+ *
+ * - 'TimeoutID' is an opaque type declared in Flow's core definition file,
+ *   used as the return type of the core setTimeout() function.
+ */
+
+export type Message = {
+  [key: string]: JSONType,
+  +type: string
+}
+
+export type PubSubClient = {
+  connectionTimeoutID: TimeoutID | void,
+  +customEventHandlers: Object,
+  failedConnectionAttempts: number,
+  isNew: boolean,
+  +listeners: Object,
+  +messageHandlers: Object,
+  nextConnectionAttemptDelayID: TimeoutID | void,
+  +options: Object,
+  +pendingSubscriptionSet: Set<string>,
+  +pendingUnsubscriptionSet: Set<string>,
+  pingTimeoutID: TimeoutID | void,
+  shouldReconnect: boolean,
+  socket: WebSocket | null,
+  +subscriptionSet: Set<string>,
+  +url: string,
+  // Methods
+  clearAllTimers(): void,
+  connect(): void,
+  destroy(): void,
+  emit(type: string, detail?: any): void,
+  pub(contractID: string, data: JSONType): void,
+  scheduleConnectionAttempt(): void,
+  sub(contractID: string): void,
+  unsub(contractID: string): void
+}
+
+export type SubMessage = {
+  [key: string]: JSONType,
+  +type: 'sub',
+  +contractID: string
+}
+
+export type UnsubMessage = {
+  [key: string]: JSONType,
+  +type: 'unsub',
+  +contractID: string
+}
+
+// ====== Enums ====== //
+
+export const NOTIFICATION_TYPE = Object.freeze({
+  ENTRY: 'entry',
+  PING: 'ping',
+  PONG: 'pong',
+  PUB: 'pub',
+  SUB: 'sub',
+  UNSUB: 'unsub'
+})
+
+export const REQUEST_TYPE = Object.freeze({
+  PUB: 'pub',
+  SUB: 'sub',
+  UNSUB: 'unsub'
+})
+
+export const RESPONSE_TYPE = Object.freeze({
+  ERROR: 'error',
+  SUCCESS: 'success'
+})
+
+export type NotificationTypeEnum = $Values<typeof NOTIFICATION_TYPE>
+export type RequestTypeEnum = $Values<typeof REQUEST_TYPE>
+export type ResponseTypeEnum = $Values<typeof RESPONSE_TYPE>
+
+// ====== API ====== //
+
+/**
+ * Creates a pubsub client instance.
+ *
+ * @param {string} url - A WebSocket URL to connect to.
+ * @param {Object?} options
+ * {boolean?} debug
+ * {object?} handlers - Custom handlers for WebSocket events.
+ * {boolean?} manual - Whether the factory should call 'connect()' automatically.
+ *   Also named 'autoConnect' or 'startClosed' in other libraries.
+ * {object?} messageHandlers - Custom handlers for different message types.
+ * {number?} pingTimeout - How long to wait for the server to send a ping, in milliseconds.
+ * {boolean?} reconnectOnDisconnection - Whether to reconnect after a server-side disconnection.
+ * {boolean?} reconnectOnOnline - Whether to reconnect after coming back online.
+ * {boolean?} reconnectOnTimeout - Whether to reconnect after a connection timeout.
+ * {number?} timeout - Connection timeout duration in milliseconds.
```

Comment:
I don't see what the default `timeout` is, it would be very helpful if you could add that here to the comments (along with the defaults for the other values as you feel is relevant)